modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -814,25 +844,106 @@ private CompletableFuture<RowId> takeLocksForUpsert(ByteBuffer searchKey, UUID i
     /**
      * Takes all required locks on a key, before inserting the value.
      *
-     * @param searchKey Key to search.
-     * @param indexId   Index id.
-     * @param txId      Transaction id.
+     * @param tableRow Table row.
+     * @param txId Transaction id.
      * @return Future completes with {@link RowId} or {@code null} if there is no value.
      */
-    private CompletableFuture<RowId> takeLocksForInsert(ByteBuffer searchKey, UUID indexId, UUID txId) {
-        return lockManager.acquire(txId, new LockKey(indexId, searchKey), LockMode.S) // Index S lock
-                .thenCompose(sharedIdxLock -> {
-                    RowId rowId = rowIdByKey(indexId, searchKey);
-
-                    if (rowId == null) {
-                        return lockManager.acquire(txId, new LockKey(indexId, searchKey), LockMode.X) // Index X lock
-                                .thenCompose(exclusiveIdxLock ->
-                                        lockManager.acquire(txId, new LockKey(tableId), LockMode.IX) // IX lock on table
-                                                .thenApply(tblLock -> null));
-                    }
+    private CompletableFuture<?> takeLocksForInsert(BinaryRow tableRow, UUID txId) {
+        return takeLockOnIndexes(tableRow, txId)
```

Comment:
Seems that it's mandatory to take locks on keys for all indexes **only** if there is at least one sorted index, because otherwise we'll lock whole table. Will clarify.

Reply:
Clarified. We need to lock keys on all indexes in any case.